build.sbt

```diff
@@ -275,3 +279,22 @@ inThisBuild(Seq(
 
 def inCompileAndTest(ss: SettingsDefinition*): Seq[Setting[_]] =
   Seq(Compile, Test) flatMap (inConfig(_)(Def.settings(ss: _*)))
+
+
+// TODO move into sbt-house-rules?
+def cleanupTests(loader: ClassLoader, log: sbt.internal.util.ManagedLogger): Unit = {
+  // shutdown Log4J to avoid classloader leaks
+  try {
+    val logManager = Class.forName("org.apache.logging.log4j.LogManager")
+    logManager.getMethod("shutdown").invoke(null)
```

Comment:
Since I am using async logging, I am guessing calling shutdown right away would risk some logs not showing up. But I think it would make sense to remove loggers that are no longer needed at some point.

Reply:
Log4J might block until these are written:

```java
    /**
     * Shutdown using the LoggerContext appropriate for the caller of this method.
     * This is equivalent to calling {@code LogManager.shutdown(false)}.
     *
     * This call is synchronous and will block until shut down is complete.
     * This may include flushing pending log events over network connections.
     *
     * @since 2.6
     */
    public static void shutdown() {
        shutdown(false);
    }
```

I haven't tested this.

I don't think it matters much for tests in any case.

Reply:
Another problem is that some appenders are long-living (relay appender used by server, backing appender, I think), so I can't stop them. If the loggers are leaking the loaders then I'd need to find out how to prevent that in general, I think.

Reply:
To be clear, this is shutting down Log4J that is in the classloader of the tests themselves, not Log4J used in the SBT build that is calling the tests.

Reply:
https://logging.apache.org/log4j/2.x/manual/webapp.html

>  Because of the nature of class loaders within web applications, Log4j resources cannot be cleaned up through normal means. Log4j must be "started" when the web application deploys and "shut down" when the web application undeploys.

Reply:
I see. Wouldn't `val logManager = Class.forName("org.apache.logging.log4j.LogManager")` load log4j in the context of the build tho?

Reply:
Oh, you're right. I meant to use `loader`. I'll push a fix.